### PR TITLE
Add `sdk/telemetry` package

### PR DESCRIPTION
**Pull request by MrAlias**
This package is the replacement for `pdata` within the `sdk` package.

It is designed to be ...

- Have minimal dependencies outside of the standard library (or anything the `otel` package already depends on)
- Be compatible with OTLP's protobuf JSON encoding
  - This means it will provide a forwards compatible binary encoding that users can rely on
- Have a reasonably user friendly Go API
- Reasonably performant

It is opinionated, and designed to work best for our needs.

Part of #954 
Part of #1106 
Split from #1126

### Follow up work

- [ ] Add `sdk/telemetry/test` pacakge to test the compatibility with `pdata`
- [ ] Update the `sdk` package to use `telemetry` as its data-model

### CODE REVIEW

File: sdk/telemetry/value.go [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+	"math"
+	"slices"
+	"strconv"
+	"unsafe"
+)
+
+// A Value represents a structured log value.
+// A zero value is valid and represents an empty value.
+type Value struct {
+	// Ensure forward compatibility by explicitly making this not comparable.
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/telemetry/value.go:25]: If this needs to be extended on the future to accommodate non-comparable changes we won't cause backwards incompatibility.

File: sdk/telemetry/span.go [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+	// dropped_links_count is the number of dropped links after the maximum size was
+	// enforced. If this value is 0, then no links were dropped.
+	DroppedLinks uint32 `json:"droppedLinksCount,omitempty"`
+	// An optional final status for this span. Semantically when Status isn't set, it means
+	// span's status code is unset, i.e. assume STATUS_CODE_UNSET (code = 0).
+	Status *Status `json:"status,omitempty"`
+}
+
+// MarshalJSON encodes s into OTLP formatted JSON.
+func (s Span) MarshalJSON() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/telemetry/span.go:125]: It is tested in the next PR I plan to add. It is a dedicated `test` package that isolates testing dependencies. The PR linked in the description has the planned changes.

File: sdk/telemetry/span.go [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+	type Alias Span
+	return json.Marshal(struct {
+		Alias
+		StartTime uint64 `json:"startTimeUnixNano,omitempty"`
+		EndTime   uint64 `json:"endTimeUnixNano,omitempty"`
+	}{
+		Alias:     Alias(s),
+		StartTime: uint64(s.StartTime.UnixNano()),
+		EndTime:   uint64(s.EndTime.UnixNano()),
+	})
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/telemetry/span.go:135]: They need to be OTLP compatible which is not the default for `time.Time`. It needs to be a string of an integer that represents the Unix timestamp in nanoseconds.

[the remark this answers is not in the document]
Comment (RonFed) [reply sdk/telemetry/span.go:135]: Got it. Does this mean that the encoded span will include two start timestamps and two end timestamps?

**Comment by RonFed** [comment]
What do you think about renaming the package to something like: `sdk/encoding`, `sdk/datamodel` or `sdk/data`?

File: sdk/telemetry/span.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/telemetry/span.go:135]: It shouldn't, there is only one `StartTime` or `EndTime` ultimately in the struct that is rendered. I can double check though.

**Comment by MrAlias** [comment]
> What do you think about renaming the package to something like: `sdk/encoding`, `sdk/datamodel` or `sdk/data`?

I am trying to stay away from things like `data` or `datamodel` as they are overly generic and don't explicate what they contain explicitly and are apt for name collisions with other "data-models".

Naming the package `encoding` was considered. However, the types themselves do not represent the actual encoding, and if it is interpreted as a verb, this would only cover part of the provided functionality (it also supports decoding). This also suffers from the "overly generic" naming issue.

I preferred using a package name that describes uniquely and clearly what the package contains. I'm open to other names, but would like something that describes `telemetry` data that is going to be represented and interacted with.

File: sdk/telemetry/span.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/telemetry/span.go:135]: Running the `TestEncode` in `test` from #1126:

```terminal
$ go test -v -run=TestEncode
=== RUN   TestEncode
    conversion_test.go:239: {"resourceSpans":[{"resource":{"attributes":[{"key":"host","value":{"stringValue":"hal"}},{"key":"id","value":{"intValue":"42"}}]},"scopeSpans":[{"scope":{"name":"go.opentelemetry.io/auto/sdk/telemetry/test","version":"v0.0.1","attributes":[{"key":"department","value":{"stringValue":"ops"}}],"droppedAttributesCount":1},"spans":[{"traceId":"01000000000000000000000000000000","spanId":"0200000000000000","traceState":"test=a","parentSpanId":"0100000000000000","name":"span.a","kind":3,"attributes":[{"key":"user","value":{"stringValue":"Alice"}},{"key":"admin","value":{"boolValue":true}},{"key":"floor","value":{"intValue":"-2"}},{"key":"impact","value":{"doubleValue":0.21362}},{"key":"reports","value":{"arrayValue":{"values":[{"stringValue":"Bob"},{"stringValue":"Dave"}]}}},{"key":"favorites","value":{"kvlistValue":{"values":[{"key":"food","value":{"stringValue":"hot dog"}},{"key":"number","value":{"intValue":"13"}}]}}},{"key":"secret","value":{"bytesValue":"TlVJNFJVWkdSamM1T0RBek9ERXdNMFF5TmpsQ05qTXpPREV6UmtNMk1FTT0="}}],"droppedAttributesCount":2,"events":[{"name":"span.event","attributes":[{"key":"impact","value":{"doubleValue":0.4372}}],"droppedAttributesCount":2,"timeUnixNano":946684800000010000}],"droppedEventsCount":3,"links":[{"traceId":"02000000000000000000000000000000","spanId":"0100000000000000","traceState":"test=green","attributes":[{"key":"queue","value":{"intValue":"17"}}],"droppedAttributesCount":8}],"droppedLinksCount":4,"status":{"message":"okay","code":1},"startTimeUnixNano":946684800000000000,"endTimeUnixNano":946684801000000000}],"schemaUrl":"http://go.opentelemetry.io/schema/v1.0.0"}],"schemaUrl":"http://go.opentelemetry.io/schema/v1.0.0"}]}

--- PASS: TestEncode (0.00s)
PASS
```

The pretty format JSON:

```
{
  "resourceSpans": [
    {
      "resource": {
        "attributes": [
          {
            "key": "host",
            "value": {
              "stringValue": "hal"
            }
          },
          {
            "key": "id",
            "value": {
              "intValue": "42"
            }
          }
        ]
      },
      "scopeSpans": [
        {
          "scope": {
            "name": "go.opentelemetry.io/auto/sdk/telemetry/test",
            "version": "v0.0.1",
            "attributes": [
              {
                "key": "department",
                "value": {
                  "stringValue": "ops"
                }
              }
            ],
            "droppedAttributesCount": 1
          },
          "spans": [
            {
              "traceId": "01000000000000000000000000000000",
              "spanId": "0200000000000000",
              "traceState": "test=a",
              "parentSpanId": "0100000000000000",
              "name": "span.a",
              "kind": 3,
              "attributes": [
                {
                  "key": "user",
                  "value": {
                    "stringValue": "Alice"
                  }
                },
                {
                  "key": "admin",
                  "value": {
                    "boolValue": true
                  }
                },
                {
                  "key": "floor",
                  "value": {
                    "intValue": "-2"
                  }
                },
                {
                  "key": "impact",
                  "value": {
                    "doubleValue": 0.21362
                  }
                },
                {
                  "key": "reports",
                  "value": {
                    "arrayValue": {
                      "values": [
                        {
                          "stringValue": "Bob"
                        },
                        {
                          "stringValue": "Dave"
                        }
                      ]
                    }
                  }
                },
                {
                  "key": "favorites",
                  "value": {
                    "kvlistValue": {
                      "values": [
                        {
                          "key": "food",
                          "value": {
                            "stringValue": "hot dog"
                          }
                        },
                        {
                          "key": "number",
                          "value": {
                            "intValue": "13"
                          }
                        }
                      ]
                    }
                  }
                },
                {
                  "key": "secret",
                  "value": {
                    "bytesValue": "TlVJNFJVWkdSamM1T0RBek9ERXdNMFF5TmpsQ05qTXpPREV6UmtNMk1FTT0="
                  }
                }
              ],
              "droppedAttributesCount": 2,
              "events": [
                {
                  "name": "span.event",
                  "attributes": [
                    {
                      "key": "impact",
                      "value": {
                        "doubleValue": 0.4372
                      }
                    }
                  ],
                  "droppedAttributesCount": 2,
                  "timeUnixNano": 946684800000010000
                }
              ],
              "droppedEventsCount": 3,
              "links": [
                {
                  "traceId": "02000000000000000000000000000000",
                  "spanId": "0100000000000000",
                  "traceState": "test=green",
                  "attributes": [
                    {
                      "key": "queue",
                      "value": {
                        "intValue": "17"
                      }
                    }
                  ],
                  "droppedAttributesCount": 8
                }
              ],
              "droppedLinksCount": 4,
              "status": {
                "message": "okay",
                "code": 1
              },
              "startTimeUnixNano": 946684800000000000,
              "endTimeUnixNano": 946684801000000000
            }
          ],
          "schemaUrl": "http://go.opentelemetry.io/schema/v1.0.0"
        }
      ],
      "schemaUrl": "http://go.opentelemetry.io/schema/v1.0.0"
    }
  ]
}
```

So, yes, with certainty, this does not duplicate the time encoding.

**Review by RonFed — APPROVED** [review]

File: sdk/telemetry/resource.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package telemetry
+
+// Resource information.
+type Resource struct {
+	// Attrs are the set of attributes that describe the resource. Attribute
+	// keys MUST be unique (it is not allowed to have more than one attribute
+	// with the same key).
+	Attrs []Attr `json:"attributes"`
```
Comment (pellared) [inline sdk/telemetry/resource.go:11]:
```suggestion
	Attrs []Attr `json:"attributes,omitempty"`
```

File: sdk/telemetry/scope.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package telemetry
+
+// Scope is the identifying values of the instrumentation scope.
+type Scope struct {
+	Name         string `json:"name,omitempty"`
+	Version      string `json:"version,omitempty"`
+	Attrs        []Attr `json:"attributes"`
```
Comment (pellared) [inline sdk/telemetry/scope.go:10]:
```suggestion
	Attrs        []Attr `json:"attributes,omitempty"`
```

File: sdk/telemetry/resource.go [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+	// DroppedAttrs is the number of dropped attributes. If the value
+	// is 0, then no attributes were dropped.
+	DroppedAttrs uint32 `json:"droppedAttributesCount,omitempty"`
```
Comment (pellared) [inline sdk/telemetry/resource.go:14]: Do we also need handling `dropped_attributes_count`?

Comment (MrAlias) [reply sdk/telemetry/resource.go:22]: Updated. PTAL.

**Review by pellared — APPROVED** [review]

**Pull request merged**